dates/parse_test.go

```diff
@@ -0,0 +1,101 @@
+package dates_test
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/nyaruka/gocommon/dates"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseDateTime(t *testing.T) {
+	kigali, err := time.LoadLocation("Africa/Kigali")
+	require.NoError(t, err)
+
+	tests := []struct {
+		layout   string
+		value    string
+		tz       *time.Location
+		expected time.Time
+		err      string
+	}{
+		{"YYYY-MM-DD t:mm", "2018-12-30 14:45", time.UTC, time.Date(2018, 12, 30, 14, 45, 0, 0, time.UTC), ""},
+		{"YYYY-MM-DD t:mm", "2018-12-30 14:45", kigali, time.Date(2018, 12, 30, 14, 45, 0, 0, kigali), ""},
+		{"YY/M/D t:mm:ss.fff", "21/1/2 9:15:30.123", kigali, time.Date(2021, 1, 2, 9, 15, 30, 123000000, kigali), ""},
+
+		{"xxx", "Mon", time.UTC, dates.ZeroDateTime, "'xxx' is not valid in a datetime format"},
+	}
+
+	for _, tc := range tests {
+		desc := fmt.Sprintf("'%s' as '%s'", tc.value, tc.layout)
+
+		actual, err := dates.ParseDateTime(tc.layout, tc.value, tc.tz)
+		if tc.err == "" {
+			assert.NoError(t, err, "unexpected error for %s", desc)
+			assert.Equal(t, tc.expected, actual, "parse mismatch for %s", desc)
+		} else {
+			assert.EqualError(t, err, tc.err, "error mismatch for %s", desc)
+			assert.Equal(t, dates.ZeroDateTime, actual)
+		}
+	}
+}
+
+func TestParseDate(t *testing.T) {
+	tests := []struct {
+		layout   string
+		value    string
+		expected dates.Date
+		err      string
+	}{
+		{"YYYY-MM-DD", "2018-12-30", dates.NewDate(2018, 12, 30), ""},
+		{"DD.MM.YYYY", "20.04.2021", dates.NewDate(2021, 4, 20), ""},
+
+		{"DD/MM/YY", "1/3/21", dates.ZeroDate, `parsing time "1/3/21" as "02/01/06": cannot parse "1/3/21" as "02"`}, // error because DD/02 requires zero-padding
```

Comment:
still pondering on how to improve these error messages since they reference parts of the format layout after it's been converted to go

Reply:
Improved - turns out time.Parse gives structured errors and we can work backwards to get the original layout element